task/backend/scheduler/treescheduler.go

```diff
@@ -0,0 +1,403 @@
+//
+// The main loop:
+//
+// The main loop waits on a time.Timer to grab the task with the minimum time.  Once it successfully grabs a task ready
+// to trigger, it will start walking the btree from the item nearest
+//
+// Putting a task into the scheduler:
+//
+// Adding a task to the scheduler acquires a write lock, grabs the task from the uniqueness map, and replaces the item
+// in the uniqueness index and btree.  If new task would trigger sooner than the current soonest triggering task, it
+// replaces the Timer when added to the scheduler.  Finally it releases the write lock.
+//
+// Removing a task from the scheduler:
+//
+// Removing a task from the scheduler acquires a write lock, deletes the task from the uniqueness index and from the
+// btree, then releases the lock.  We do not have to readjust the time on delete, because, if the minimum task isn't
+// ready yet, the main loop just resets the timer and keeps going.
+type TreeScheduler struct {
+	sync.RWMutex
+	scheduled    *btree.BTree
+	nextTime     map[ID]ordering // we need this index so we can delete items from the scheduled
+	when         time.Time
+	executor     Executor
+	onErr        ErrorFunc
+	time         Time
+	timer        Timer
+	done         chan struct{}
+	workchans    []chan Item
+	wg           sync.WaitGroup
+	checkpointer SchedulableService
+
+	sm *SchedulerMetrics
+}
+
+type ExecutorFunc func(ctx context.Context, id ID, scheduledAt time.Time) error
+
+type ErrorFunc func(ctx context.Context, taskID ID, scheduledAt time.Time, err error)
+
+type treeSchedulerOptFunc func(t *TreeScheduler) error
+
+func WithOnErrorFn(fn ErrorFunc) treeSchedulerOptFunc {
+	return func(t *TreeScheduler) error {
+		t.onErr = fn
+		return nil
+	}
+}
+
+func WithMaxConcurrentWorkers(n int) treeSchedulerOptFunc {
+	return func(t *TreeScheduler) error {
+		t.workchans = make([]chan Item, n)
+		return nil
+	}
+}
+
+func WithTime(t Time) treeSchedulerOptFunc {
+	return func(sch *TreeScheduler) error {
+		sch.time = t
+		return nil
+	}
+}
+
+// Executor is any function that accepts an ID, a time, and a duration.
+func NewScheduler(executor Executor, checkpointer SchedulableService, opts ...treeSchedulerOptFunc) (*TreeScheduler, *SchedulerMetrics, error) {
+	s := &TreeScheduler{
+		executor:     executor,
+		scheduled:    btree.New(degreeBtreeScheduled),
+		nextTime:     map[ID]ordering{},
+		onErr:        func(_ context.Context, _ ID, _ time.Time, _ error) {},
+		time:         stdTime{},
+		done:         make(chan struct{}, 1),
+		checkpointer: checkpointer,
+	}
+
+	// apply options
+	for i := range opts {
+		if err := opts[i](s); err != nil {
+			return nil, nil, err
+		}
+	}
+	if s.workchans == nil {
+		s.workchans = make([]chan Item, defaultMaxWorkers)
+
+	}
+
+	s.wg.Add(len(s.workchans))
+	for i := 0; i < len(s.workchans); i++ {
+		s.workchans[i] = make(chan Item)
+		go s.work(i)
```

Comment:
I would rather see this call take a chan than an int. It is hard to follow what the int _means_. If for any reason we wanted to adjust the number of workers passing the index becomes kinda brittle.